File: cpp/dynamicProgramming/countPaths.cpp
#include <iostream>
#include <vector>

using std::cout;
using std::endl;
using std::vector;

// Brute Force - Time: O(2 ^ (n + m)), Space: O(n + m)
int bruteForce(int r, int c, int rows, int cols) {
    if (r == rows or c == cols) {
        return 0;
    }
    if (r == rows - 1 or c == cols - 1) {
        return 1;
    }
    return bruteForce(r + 1, c, rows, cols) +
        bruteForce(r, c + 1, rows, cols);
}

// Memoization - Time and Space: O(n * m)
int memoization(int r, int c, int rows, int cols, vector<vector<int>>& cache) {
    if (r == rows or c == cols) {
        return 0;
    }
    if (cache[r][c] > 0) {
        return cache[r][c];
    }
    if (r == rows - 1 or c == cols - 1) {
        return 1;
    }
    cache[r][c] = memoization(r + 1, c, rows, cols, cache) +  
        memoization(r, c + 1, rows, cols, cache);
    return cache[r][c];
}

// Dynamic Programming - Time: O(n * m), Space: O(m), where m is num of cols
int dp(int rows, int cols) {
    int *prevRow = new int[cols]();

    for (int r = rows - 1; r >= 0; r--) {
        int* curRow = new int[cols]();
        curRow[cols - 1] = 1;
        for (int c = cols - 2; c >= 0; c--) {
            curRow[c] = curRow[c + 1] + prevRow[c];
        }
        prevRow = curRow;
    }
    return prevRow[0];
}

int main() {
    // F(0) = 0
    // F(1) = 1
    // F(2) = 1
    // F(3) = 2
    // F(4) = 3
    // F(5) = 5
    cout << bruteForce(0, 0, 4, 4) << endl;
    vector<vector<int>> cache(4, vector<int>(4));
    //int **cache = new int [4][4]();
    // for (int i = 0; i < 4; i++) {
    //     for (int j = 0; i < 4; j++) {
    //         cache[i][j] = 0;
    //     }
    // }
    cout << memoization(0, 0, 4, 4, cache) << endl;
    cout << dp(4, 4) << endl;
    return 0;
}

File: csharp/dynamicProgramming/CountPaths.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NeetCodeTester.dynamicProgramming
{
    class CountPaths
    {

        // Brute Force - Time: O(2 ^ (n + m)), Space: O(n + m)
        public static int BruteForce(int r, int c, int rows, int cols)
        {
            if (r == rows || c == cols)
            {
                return 0;
            }
            if (r == rows - 1 && c == cols - 1)
            {
                return 1;
            }
            return (BruteForce(r + 1, c, rows, cols) +
                    BruteForce(r, c + 1, rows, cols));
        }

        // Memoization - Time and Space: O(n * m)
        public static int Memoization(int r, int c, int rows, int cols, int[][] cache)
        {
            if (r == rows || c == cols)
            {
                return 0;
            }
            if (cache[r][c] > 0)
            {
                return cache[r][c];
            }
            if (r == rows - 1 && c == cols - 1)
            {
                return 1;
            }
            cache[r][c] = (Memoization(r + 1, c, rows, cols, cache) +
                            Memoization(r, c + 1, rows, cols, cache));
            return cache[r][c];
        }

        // Dynamic Programming - Time: O(n * m), Space: O(m), where m is num of cols
        public static int Dp(int rows, int cols)
        {
            int[] prevRow = new int[cols];

            for (int i = rows - 1; i >= 0; i--)
            {
                int[] curRow = new int[cols];
                curRow[cols - 1] = 1;
                for (int j = cols - 2; j >= 0; j--)
                {
                    curRow[j] = curRow[j + 1] + prevRow[j];
                }
                prevRow = curRow;
            }
            return prevRow[0];
        }
    }
}

 